File: backend/src/routes/orgs.ts
import type { FastifyInstance } from "fastify";
import { query, withTransaction } from "../db";
import { ConflictError, ForbiddenError, NotFoundError } from "../errors";
import { getCurrentMembership, getCurrentUser, requireCurrentUser, requireRole } from "../middleware/auth";
import { inviteSchema, memberRoleSchema, orgCreateSchema, orgUpdateSchema } from "../schemas/org";
import {
  SQL_CHECK_ORG_SLUG_EXISTS,
  SQL_DELETE_MEMBERSHIP,
  SQL_INSERT_MEMBERSHIP,
  SQL_INSERT_ORG,
  SQL_SELECT_MEMBERS,
  SQL_SELECT_ORGS_FOR_USER,
  SQL_SELECT_PENDING_INVITATIONS,
  SQL_SELECT_USER_BY_ID,
  SQL_UPDATE_MEMBERSHIP_ROLE,
  buildUpdateOrgSql,
  getOrgBySlug,
} from "../sql/orgs";
import { acceptInvitation, createInvitation } from "../services/invitation-service";
import { mapInvitation, mapMembership, mapOrg } from "../utils/serialization";

export default async function orgRoutes(api: FastifyInstance) {
  api.get("/orgs", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const result = await query(SQL_SELECT_ORGS_FOR_USER, [currentUser.id]);
    return result.rows.map(mapOrg);
  });

  api.post("/orgs", { preHandler: requireCurrentUser }, async (request, reply) => {
    const payload = orgCreateSchema.parse(request.body);
    const currentUser = getCurrentUser(request);
    const result = await withTransaction(async (client) => {
      const existing = await client.query(SQL_CHECK_ORG_SLUG_EXISTS, [payload.slug]);
      if (existing.rowCount) throw new ConflictError("Slug already taken");
      const orgResult = await client.query(SQL_INSERT_ORG, [payload.slug, payload.name]);
      await client.query(SQL_INSERT_MEMBERSHIP, [orgResult.rows[0].id, currentUser.id, "owner"]);
      return orgResult.rows[0];
    });
    reply.status(201).send(mapOrg(result));
  });

  api.get("/orgs/:slug", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { slug } = request.params as { slug: string };
    await getCurrentMembership(slug, currentUser.id);
    return mapOrg(await getOrgBySlug(slug));
  });

  api.patch("/orgs/:slug", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { slug } = request.params as { slug: string };
    const payload = orgUpdateSchema.parse(request.body);
    const membership = await getCurrentMembership(slug, currentUser.id);
    requireRole(membership, ["owner", "admin"]);
    const sets: string[] = [];
    const values: unknown[] = [slug];
    if ("name" in payload) {
      values.push(payload.name ?? null);
      sets.push(`name = $${values.length}`);
    }
    if ("settings" in payload) {
      values.push(payload.settings ?? null);
      sets.push(`settings = $${values.length}`);
    }
    if (!sets.length) return mapOrg(await getOrgBySlug(slug));
    const result = await query(buildUpdateOrgSql(sets), values);
    return mapOrg(result.rows[0]);
  });

  api.get("/orgs/:slug/members", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { slug } = request.params as { slug: string };
    const org = await getOrgBySlug(slug);
    await getCurrentMembership(slug, currentUser.id);
    const result = await query(SQL_SELECT_MEMBERS, [org.id]);
    return result.rows.map((row) =>
      mapMembership(row, {
        id: row.user_id_ref,
        email: row.email,
        full_name: row.full_name,
        avatar_url: row.avatar_url,
      })
    );
  });

  api.delete("/orgs/:slug/members/:userId", { preHandler: requireCurrentUser }, async (request, reply) => {
    const currentUser = getCurrentUser(request);
    const { slug, userId } = request.params as { slug: string; userId: string };
    const org = await getOrgBySlug(slug);
    const membership = await getCurrentMembership(slug, currentUser.id);
    requireRole(membership, ["owner", "admin"]);
    if (currentUser.id === userId) throw new ForbiddenError("Cannot remove yourself");
    const result = await query(SQL_DELETE_MEMBERSHIP, [org.id, userId]);
    if (!result.rowCount) throw new NotFoundError("Member not found");
    reply.status(204).send();
  });

  api.patch("/orgs/:slug/members/:userId", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { slug, userId } = request.params as { slug: string; userId: string };
    const payload = memberRoleSchema.parse(request.body);
    const org = await getOrgBySlug(slug);
    const membership = await getCurrentMembership(slug, currentUser.id);
    requireRole(membership, ["owner"]);
    const result = await query(SQL_UPDATE_MEMBERSHIP_ROLE, [org.id, userId, payload.role]);
    if (!result.rowCount) throw new NotFoundError("Member not found");
    const userResult = await query(SQL_SELECT_USER_BY_ID, [userId]);
    return mapMembership(result.rows[0], userResult.rows[0]);
  });

  api.post("/orgs/:slug/invitations", { preHandler: requireCurrentUser }, async (request, reply) => {
    const currentUser = getCurrentUser(request);
    const { slug } = request.params as { slug: string };
    const payload = inviteSchema.parse(request.body);
    const org = await getOrgBySlug(slug);
    const membership = await getCurrentMembership(slug, currentUser.id);
    requireRole(membership, ["owner", "admin"]);
    const invitation = await withTransaction((client) =>
      createInvitation(client, org.id, currentUser.id, payload.email, payload.role)
    );
    reply.status(201).send(mapInvitation(invitation));
  });

  api.get("/orgs/:slug/invitations", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { slug } = request.params as { slug: string };
    const org = await getOrgBySlug(slug);
    const membership = await getCurrentMembership(slug, currentUser.id);
    requireRole(membership, ["owner", "admin"]);
    const result = await query(SQL_SELECT_PENDING_INVITATIONS, [org.id]);
    return result.rows.map(mapInvitation);
  });

  api.post("/invitations/:token/accept", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { token } = request.params as { token: string };
    const membership = await withTransaction((client) => acceptInvitation(client, token, currentUser));
    const userResult = await query(SQL_SELECT_USER_BY_ID, [membership.user_id]);
    return mapMembership(membership, userResult.rows[0]);
  });
}

File: backend/app/routes/orgs.py
from __future__ import annotations

import re
from typing import Any

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from sqlalchemy import func, update

import app.sql.orgs as sql_orgs
from app.db.pool import get_pool
from app.db.tables import organizations
from app.errors import NotFoundError
from app.middleware.auth import get_current_membership, get_current_user, require_role
from app.services.invitation_service import accept_invitation, create_invitation

router = APIRouter()

_ALLOWED_ORG_PATCH = {"name", "plan", "settings"}


def _map_org(row: Any) -> dict:
    return {
        "id": str(row["id"]),
        "slug": row["slug"],
        "name": row["name"],
        "plan": row.get("plan"),
        "settings": row.get("settings"),
        "created_at": row["created_at"].isoformat() if row.get("created_at") else None,
    }


def _map_member(row: Any) -> dict:
    return {
        "id": str(row["id"]),
        "org_id": str(row["org_id"]),
        "user_id": str(row["user_id"]),
        "role": row["role"],
        "joined_at": row["joined_at"].isoformat() if row.get("joined_at") else None,
        "user": {
            "id": str(row["user_id"]),
            "email": row.get("email"),
            "full_name": row.get("full_name"),
            "avatar_url": row.get("avatar_url"),
        },
    }


def _map_invitation(row: Any) -> dict:
    return {
        "id": str(row["id"]),
        "org_id": str(row["org_id"]),
        "invited_by": str(row["invited_by"]) if row.get("invited_by") else None,
        "email": row["email"],
        "role": row["role"],
        "token": row.get("token"),
        "accepted_at": row["accepted_at"].isoformat() if row.get("accepted_at") else None,
        "expires_at": row["expires_at"].isoformat() if row.get("expires_at") else None,
        "created_at": row["created_at"].isoformat() if row.get("created_at") else None,
    }


@router.get("/orgs")
async def orgs_list(current_user: dict = Depends(get_current_user)) -> list:
    pool = await get_pool()
    rows = await pool.fetch(sql_orgs.select_orgs_for_user(current_user["id"]))
    return [_map_org(r) for r in rows]


@router.post("/orgs", status_code=201)
async def orgs_create(
    request: Request, current_user: dict = Depends(get_current_user)
) -> JSONResponse:
    body = await request.json()

    name = body.get("name", "").strip()
    if not name:
        return JSONResponse({"detail": "name is required"}, status_code=400)

    def _slugify(v: str) -> str:
        slug = re.sub(r"[^a-z0-9]+", "-", v.lower()).strip("-")
        return slug[:50] or "org"

    pool = await get_pool()
    async with pool.acquire() as conn:
        base_slug = _slugify(name)
        slug = base_slug
        counter = 1
        while True:
            exists = await conn.fetchrow(sql_orgs.check_org_slug_exists(slug))
            if exists is None:
                break
            slug = f"{base_slug}-{counter}"
            counter += 1

        org_row = await conn.fetchrow(sql_orgs.insert_org(slug, name))
        await conn.execute(
            sql_orgs.insert_membership(org_row["id"], current_user["id"], "owner")
        )

    return JSONResponse(_map_org(org_row), status_code=201)


@router.get("/orgs/{slug}")
async def orgs_get(slug: str, current_user: dict = Depends(get_current_user)) -> dict:
    pool = await get_pool()
    row = await pool.fetchrow(sql_orgs.select_org_by_slug(slug))
    if row is None:
        raise NotFoundError("Organization not found")
    # Ensure user is a member
    await get_current_membership(slug, str(current_user["id"]))
    return _map_org(row)


@router.patch("/orgs/{slug}")
async def orgs_update(
    slug: str, request: Request, current_user: dict = Depends(get_current_user)
) -> dict:
    body = await request.json()
    membership = await get_current_membership(slug, str(current_user["id"]))
    require_role(membership, ["owner", "admin"])

    updates = {k: v for k, v in body.items() if k in _ALLOWED_ORG_PATCH}
    pool = await get_pool()
    if not updates:
        row = await pool.fetchrow(sql_orgs.select_org_by_slug(slug))
        return _map_org(row)

    stmt = (
        update(organizations)
        .where(organizations.c.slug == slug)
        .values(**updates, updated_at=func.now())
        .returning(
            organizations.c.id,
            organizations.c.slug,
            organizations.c.name,
            organizations.c.plan,
            organizations.c.settings,
            organizations.c.created_at,
        )
    )
    row = await pool.fetchrow(stmt)
    return _map_org(row)


@router.get("/orgs/{slug}/members")
async def org_members_list(
    slug: str, current_user: dict = Depends(get_current_user)
) -> list:
    membership = await get_current_membership(slug, str(current_user["id"]))
    pool = await get_pool()
    rows = await pool.fetch(sql_orgs.select_members(membership["org_id"]))
    return [_map_member(r) for r in rows]


@router.delete("/orgs/{slug}/members/{user_id}", status_code=204)
async def org_members_delete(
    slug: str,
    user_id: str,
    current_user: dict = Depends(get_current_user),
) -> None:
    membership = await get_current_membership(slug, str(current_user["id"]))
    require_role(membership, ["owner", "admin"])
    pool = await get_pool()
    deleted = await pool.fetchrow(sql_orgs.delete_membership(membership["org_id"], user_id))
    if deleted is None:
        raise NotFoundError("Member not found")


@router.patch("/orgs/{slug}/members/{user_id}")
async def org_members_update(
    slug: str,
    user_id: str,
    request: Request,
    current_user: dict = Depends(get_current_user),
) -> dict:
    body = await request.json()
    membership = await get_current_membership(slug, str(current_user["id"]))
    require_role(membership, ["owner"])
    role = body.get("role")
    if not role:
        return JSONResponse({"detail": "role is required"}, status_code=400)
    pool = await get_pool()
    row = await pool.fetchrow(
        sql_orgs.update_membership_role(membership["org_id"], user_id, role)
    )
    if row is None:
        raise NotFoundError("Member not found")
    return {
        "id": str(row["id"]),
        "org_id": str(row["org_id"]),
        "user_id": str(row["user_id"]),
        "role": row["role"],
        "joined_at": row["joined_at"].isoformat() if row.get("joined_at") else None,
    }


@router.post("/orgs/{slug}/invitations", status_code=201)
async def org_invitations_create(
    slug: str,
    request: Request,
    current_user: dict = Depends(get_current_user),
) -> JSONResponse:
    body = await request.json()
    membership = await get_current_membership(slug, str(current_user["id"]))
    require_role(membership, ["owner", "admin"])
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await create_invitation(
            conn,
            org_id=str(membership["org_id"]),
            invited_by=str(current_user["id"]),
            email=body["email"],
            role=body.get("role", "member"),
        )
    return JSONResponse(_map_invitation(row), status_code=201)


@router.get("/orgs/{slug}/invitations")
async def org_invitations_list(
    slug: str, current_user: dict = Depends(get_current_user)
) -> list:
    membership = await get_current_membership(slug, str(current_user["id"]))
    pool = await get_pool()
    rows = await pool.fetch(sql_orgs.select_pending_invitations(membership["org_id"]))
    return [_map_invitation(r) for r in rows]


@router.post("/invitations/{token}/accept")
async def invitation_accept(
    token: str, current_user: dict = Depends(get_current_user)
) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        result = await accept_invitation(conn, token, current_user)
    return {
        "id": str(result["id"]),
        "org_id": str(result["org_id"]),
        "user_id": str(result["user_id"]),
        "role": result["role"],
    }
